cfme/tests/generic_objects/test_instances.py

```diff
@@ -56,3 +115,127 @@ def test_generic_objects_crud(appliance, context, request):
+                name='rest_generic_instance{}'.format(fauxfactory.gen_alphanumeric()),
+                definition=definition,
+                attributes={'addr01': 'Test Address'}
+            )
+            request.addfinalizer(instance.delete)
+        view = navigate_to(instance, 'Details')
+        assert view.is_displayed
+
+
+@pytest.mark.parametrize('button_group', [True, False],
+                         ids=['button_group_with_button', 'single_button'])
+@pytest.mark.parametrize('context', [ViaUI])
+def test_generic_objects_with_buttons_ui(appliance, definition, service, request, button_group,
+                                         context):
+    """
+        Tests buttons ui visibility assigned to generic object
+
+        Metadata:
+            test_flag: ui
+    """
+    with appliance.context.use(context):
+        myservice = MyService(appliance, name=service.name)
+        group_name = 'button_group_{}'.format(fauxfactory.gen_alphanumeric())
+        view = appliance.browser.create_view(BaseLoggedInPage)
+        if button_group:
+            definition.add_button_group(
+                name=group_name,
+                description='Group_button_description',
+                image='fa-user'
+            )
+            view.flash.assert_message(
+                'Custom Button Group "{}" has been successfully added.'.format(group_name))
+
+        button_name = 'button_{}'.format(fauxfactory.gen_alphanumeric())
+        definition.add_button(
+            name=button_name,
+            description='Button_description',
+            image='fa-home',
+            request=fauxfactory.gen_alphanumeric(),
+            button_group=group_name if button_group else None)
+        message = (
+            'Custom Button "{}" has been successfully added under the selected button group.'
+            if button_group else 'Custom Button "{}" has been successfully added.'
+        )
+        view.flash.assert_message(message.format(button_name))
+    with appliance.context.use(ViaREST):
+        instance = appliance.collections.generic_objects.create(
+            name='rest_generic_instance{}'.format(fauxfactory.gen_alphanumeric()),
+            definition=definition,
+            attributes={'addr01': 'Test Address'},
+            associations={'services': [myservice]}
+        )
+        request.addfinalizer(instance.delete)
+        service.action.add_resource(
+            resource=appliance.rest_api.collections.generic_objects.find_by(
+                name=instance.name)[0]._ref_repr()
+        )
+        assert_response(appliance)
+
+    with appliance.context.use(context):
+        view = navigate_to(myservice, 'GenericObjectInstance', instance_name=instance.name)
+        if button_group:
+            assert view.toolbar.group(group_name).custom_button.has_item(button_name)
+        else:
+            assert view.toolbar.button(button_name).custom_button.is_displayed
```

Comment:
**REQUIRED**
This test is quite big, consider to move some of its parts to fixtures and leave only those actions that related to the test "Tests buttons ui visibility assigned to generic object".